File: llm-harness-js/src/stats/data-sources/arena-ai.ts
const ARENA_AI_BASE_URL = "https://arena.ai/leaderboard/text-to-image";
const REQUEST_TIMEOUT_MS = 30_000;
const MIN_VALID_ROWS = 20;
const MIN_VALID_CATEGORIES = 4;

export const ARENA_AI_DEFAULT_CATEGORY_SLUGS = [
  "commercial-design",
  "3d-modeling",
  "cartoon",
  "photorealistic",
  "art",
  "portraits",
  "text-rendering",
] as const;

type ArenaAiGroupedCategoryName =
  | "photorealistic"
  | "illustrative"
  | "contextual";

export const ARENA_AI_GROUPED_CATEGORY_SLUGS: Record<
  ArenaAiGroupedCategoryName,
  readonly string[]
> = {
  photorealistic: ["photorealistic", "portraits"],
  illustrative: ["cartoon", "art"],
  contextual: ["commercial-design", "3d-modeling", "text-rendering"],
};

const ARENA_AI_GROUPED_CATEGORY_NAMES: readonly ArenaAiGroupedCategoryName[] = [
  "photorealistic",
  "illustrative",
  "contextual",
];

const ARENA_AI_GROUP_BY_SLUG = new Map<string, ArenaAiGroupedCategoryName>(
  ARENA_AI_GROUPED_CATEGORY_NAMES.flatMap((groupName) =>
    ARENA_AI_GROUPED_CATEGORY_SLUGS[groupName].map((slug) => [slug, groupName]),
  ),
);

type NumberOrNull = number | null;

export type ArenaAiRow = {
  model: string;
  provider: string | null;
  score: NumberOrNull;
  ci95: string | null;
  votes: NumberOrNull;
};

export type ArenaAiCategoryPayload = {
  fetched_at_epoch_seconds: number;
  category_slug: string;
  source_url: string;
  final_url: string | null;
  status_code: number | null;
  challenge_detected: boolean;
  rows_with_score: number;
  rows: ArenaAiRow[];
};

export type ArenaAiAggregatedCategoryRow = {
  rank: number;
  score: NumberOrNull;
  ci95: string | null;
  votes: NumberOrNull;
};

export type ArenaAiAggregatedModel = {
  model: string;
  provider: string | null;
  category_count: number;
  average_score: NumberOrNull;
  vote_weighted_score: NumberOrNull;
  average_rank: NumberOrNull;
  votes_sum: number;
  categories: Record<string, ArenaAiAggregatedCategoryRow>;
  weighted_scores: {
    photorealistic: NumberOrNull;
    illustrative: NumberOrNull;
    contextual: NumberOrNull;
    grouped_overall: NumberOrNull;
  };
  grouped_votes: {
    photorealistic: number;
    illustrative: number;
    contextual: number;
    total: number;
  };
};

/**
 * Arena text-to-image aggregated response.
 *
 * This source is failure-safe by design and returns an empty payload when all
 * category fetches fail.
 */
export type ArenaAiOutputPayload = {
  fetched_at_epoch_seconds: number;
  base_url: string;
  category_slugs: string[];
  categories: ArenaAiCategoryPayload[];
  grouped_category_slugs: Record<ArenaAiGroupedCategoryName, readonly string[]>;
  valid_categories: string[];
  total_valid_categories: number;
  total_models_aggregated: number;
  scrape_feasible_now: boolean;
  rows: ArenaAiAggregatedModel[];
};

/**
 * Arena source options.
 */
export type ArenaAiOptions = {
  categorySlugs?: string[];
  minValidRows?: number;
  minValidCategories?: number;
};

type ArenaAiAggregateAccumulator = {
  model: string;
  provider: string | null;
  category_rows: Record<string, ArenaAiAggregatedCategoryRow>;
  category_count: number;
  votes_sum: number;
  score_sum: number;
  score_weighted_sum: number;
  rank_sum: number;
};

type ArenaAiGroupedAccumulatorValue = {
  score_weighted_sum: number;
  votes_sum: number;
  score_sum: number;
  count: number;
};

type ArenaAiGroupedAccumulator = Record<
  ArenaAiGroupedCategoryName,
  ArenaAiGroupedAccumulatorValue
>;

function nowEpochSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

function detectChallenge(html: string): boolean {
  return /challenge-platform|__CF\$cv\$params|Verify you are human|Security Verification/i.test(
    html,
  );
}

function asFiniteNumber(value: unknown): NumberOrNull {
  const numeric = Number(value);
  return Number.isFinite(numeric) ? numeric : null;
}

function extractLeaderboardRows(html: string): ArenaAiRow[] {
  const titleMatches = [...html.matchAll(/<a[^>]*\btitle="([^"]+)"/g)];
  const rows: ArenaAiRow[] = [];

  for (let index = 0; index < titleMatches.length; index += 1) {
    const title = titleMatches[index]?.[1] ?? "";
    const start = titleMatches[index]?.index ?? 0;
    const end =
      titleMatches[index + 1]?.index ?? Math.min(start + 4_000, html.length);
    const rowBlock = html.slice(start, end);
    const scoreMatch = rowBlock.match(
      />(\d{3,4})<\/span><span[^>]*>±(\d+)<\/span>/,
    );
    const votesMatch = rowBlock.match(
      /±\d+<\/span>(?:.|\n){0,1200}?>([\d,]{2,})<\/span>/,
    );
    const providerMatch = rowBlock.match(/>([^<]+ · [^<]+)<\/span>/);

    rows.push({
      model: title,
      provider: providerMatch?.[1] ?? null,
      score: asFiniteNumber(scoreMatch?.[1]),
      ci95: scoreMatch?.[2] ? `±${scoreMatch[2]}` : null,
      votes: votesMatch?.[1]
        ? asFiniteNumber(votesMatch[1].replaceAll(",", ""))
        : null,
    });
  }

  return rows;
}

async function fetchCategory(
  baseUrl: string,
  categorySlug: string,
): Promise<ArenaAiCategoryPayload> {
  const sourceUrl = `${baseUrl}/${categorySlug}`;
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
    const response = await fetch(sourceUrl, {
      headers: {
        "user-agent":
          "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Safari/537.36",
        "accept-language": "en-US,en;q=0.9",
      },
      redirect: "follow",
      signal: controller.signal,
    });
    clearTimeout(timeout);

    const html = await response.text();
    const rowsWithScore = extractLeaderboardRows(html).filter(
      (row) => row.score != null,
    );

    return {
      fetched_at_epoch_seconds: nowEpochSeconds(),
      category_slug: categorySlug,
      source_url: sourceUrl,
      final_url: response.url,
      status_code: response.status,
      challenge_detected: detectChallenge(html),
      rows_with_score: rowsWithScore.length,
      rows: rowsWithScore,
    };
  } catch {
    return {
      fetched_at_epoch_seconds: nowEpochSeconds(),
      category_slug: categorySlug,
      source_url: sourceUrl,
      final_url: null,
      status_code: null,
      challenge_detected: false,
      rows_with_score: 0,
      rows: [],
    };
  }
}

function roundTo4(value: number): number {
  return Number(value.toFixed(4));
}

function weightedScoreOrAverage(
  weightedSum: number,
  votesSum: number,
  scoreSum: number,
  count: number,
): NumberOrNull {
  if (votesSum > 0) {
    return roundTo4(weightedSum / votesSum);
  }
  if (count > 0) {
    return roundTo4(scoreSum / count);
  }
  return null;
}

function detectGroupedCategory(
  categorySlug: string,
): ArenaAiGroupedCategoryName | null {
  return ARENA_AI_GROUP_BY_SLUG.get(categorySlug) ?? null;
}

function createGroupedAccumulator(): ArenaAiGroupedAccumulator {
  return {
    photorealistic: {
      score_weighted_sum: 0,
      votes_sum: 0,
      score_sum: 0,
      count: 0,
    },
    illustrative: {
      score_weighted_sum: 0,
      votes_sum: 0,
      score_sum: 0,
      count: 0,
    },
    contextual: {
      score_weighted_sum: 0,
      votes_sum: 0,
      score_sum: 0,
      count: 0,
    },
  };
}

function buildGroupedScores(
  categoryRows: Record<string, ArenaAiAggregatedCategoryRow>,
): Pick<ArenaAiAggregatedModel, "weighted_scores" | "grouped_votes"> {
  const groupedAccumulator = createGroupedAccumulator();
  for (const [categorySlug, row] of Object.entries(categoryRows)) {
    const groupedCategory = detectGroupedCategory(categorySlug);
    if (!groupedCategory) {
      continue;
    }
    const score = row.score ?? 0;
    const votes = row.votes ?? 0;
    groupedAccumulator[groupedCategory].score_weighted_sum += score * votes;
    groupedAccumulator[groupedCategory].votes_sum += votes;
    groupedAccumulator[groupedCategory].score_sum += score;
    groupedAccumulator[groupedCategory].count += 1;
  }

  const groupedScores = {
    photorealistic: weightedScoreOrAverage(
      groupedAccumulator.photorealistic.score_weighted_sum,
      groupedAccumulator.photorealistic.votes_sum,
      groupedAccumulator.photorealistic.score_sum,
      groupedAccumulator.photorealistic.count,
    ),
    illustrative: weightedScoreOrAverage(
      groupedAccumulator.illustrative.score_weighted_sum,
      groupedAccumulator.illustrative.votes_sum,
      groupedAccumulator.illustrative.score_sum,
      groupedAccumulator.illustrative.count,
    ),
    contextual: weightedScoreOrAverage(
      groupedAccumulator.contextual.score_weighted_sum,
      groupedAccumulator.contextual.votes_sum,
      groupedAccumulator.contextual.score_sum,
      groupedAccumulator.contextual.count,
    ),
    grouped_overall: null as NumberOrNull,
  };
  const groupedTotalVotes = ARENA_AI_GROUPED_CATEGORY_NAMES.reduce(
    (sum, groupName) => sum + groupedAccumulator[groupName].votes_sum,
    0,
  );
  groupedScores.grouped_overall = weightedScoreOrAverage(
    ARENA_AI_GROUPED_CATEGORY_NAMES.reduce(
      (sum, groupName) =>
        sum + groupedAccumulator[groupName].score_weighted_sum,
      0,
    ),
    groupedTotalVotes,
    ARENA_AI_GROUPED_CATEGORY_NAMES.reduce(
      (sum, groupName) => sum + groupedAccumulator[groupName].score_sum,
      0,
    ),
    ARENA_AI_GROUPED_CATEGORY_NAMES.reduce(
      (sum, groupName) => sum + groupedAccumulator[groupName].count,
      0,
    ),
  );

  return {
    weighted_scores: groupedScores,
    grouped_votes: {
      photorealistic: groupedAccumulator.photorealistic.votes_sum,
      illustrative: groupedAccumulator.illustrative.votes_sum,
      contextual: groupedAccumulator.contextual.votes_sum,
      total: groupedTotalVotes,
    },
  };
}

function buildAggregatedRows(
  categoryPayloads: ArenaAiCategoryPayload[],
): ArenaAiAggregatedModel[] {
  const byModel = new Map<string, ArenaAiAggregateAccumulator>();

  for (const payload of categoryPayloads) {
    for (let index = 0; index < payload.rows.length; index += 1) {
      const row = payload.rows[index];
      if (row == null) {
        continue;
      }
      const existing = byModel.get(row.model);
      const aggregate = existing ?? {
        model: row.model,
        provider: row.provider,
        category_rows: {},
        category_count: 0,
        votes_sum: 0,
        score_sum: 0,
        score_weighted_sum: 0,
        rank_sum: 0,
      };

      aggregate.category_rows[payload.category_slug] = {
        rank: index + 1,
        score: row.score,
        ci95: row.ci95,
        votes: row.votes,
      };
      aggregate.category_count += 1;
      aggregate.rank_sum += index + 1;
      aggregate.score_sum += row.score ?? 0;
      const votes = row.votes ?? 0;
      aggregate.votes_sum += votes;
      aggregate.score_weighted_sum += (row.score ?? 0) * votes;

      if (!existing) {
        byModel.set(row.model, aggregate);
      }
    }
  }

  const aggregatedRows: ArenaAiAggregatedModel[] = [];
  for (const aggregate of byModel.values()) {
    const averageScore =
      aggregate.category_count > 0
        ? roundTo4(aggregate.score_sum / aggregate.category_count)
        : null;
    const voteWeightedScore =
      aggregate.votes_sum > 0
        ? roundTo4(aggregate.score_weighted_sum / aggregate.votes_sum)
        : averageScore;
    const averageRank =
      aggregate.category_count > 0
        ? roundTo4(aggregate.rank_sum / aggregate.category_count)
        : null;
    const grouped = buildGroupedScores(aggregate.category_rows);

    aggregatedRows.push({
      model: aggregate.model,
      provider: aggregate.provider,
      category_count: aggregate.category_count,
      average_score: averageScore,
      vote_weighted_score: voteWeightedScore,
      average_rank: averageRank,
      votes_sum: aggregate.votes_sum,
      categories: aggregate.category_rows,
      weighted_scores: grouped.weighted_scores,
      grouped_votes: grouped.grouped_votes,
    });
  }

  aggregatedRows.sort((left, right) => {
    const leftScore = left.vote_weighted_score;
    const rightScore = right.vote_weighted_score;
    if (leftScore != null && rightScore != null && rightScore !== leftScore) {
      return rightScore - leftScore;
    }
    if (leftScore == null && rightScore != null) {
      return 1;
    }
    if (leftScore != null && rightScore == null) {
      return -1;
    }
    return right.category_count - left.category_count;
  });

  return aggregatedRows;
}

/**
 * Fetch and aggregate Arena text-to-image leaderboard categories.
 */
export async function getArenaAiTextToImageStats(
  options: ArenaAiOptions = {},
): Promise<ArenaAiOutputPayload> {
  const categorySlugs =
    options.categorySlugs != null && options.categorySlugs.length > 0
      ? options.categorySlugs
      : [...ARENA_AI_DEFAULT_CATEGORY_SLUGS];
  const minValidRows = options.minValidRows ?? MIN_VALID_ROWS;
  const minValidCategories = options.minValidCategories ?? MIN_VALID_CATEGORIES;

  const categories = await Promise.all(
    categorySlugs.map((slug) => fetchCategory(ARENA_AI_BASE_URL, slug)),
  );
  const validCategories = categories.filter(
    (category) => category.rows_with_score >= minValidRows,
  );
  const rows = buildAggregatedRows(validCategories);

  return {
    fetched_at_epoch_seconds: nowEpochSeconds(),
    base_url: ARENA_AI_BASE_URL,
    category_slugs: categorySlugs,
    categories,
    grouped_category_slugs: ARENA_AI_GROUPED_CATEGORY_SLUGS,
    valid_categories: validCategories.map((category) => category.category_slug),
    total_valid_categories: validCategories.length,
    total_models_aggregated: rows.length,
    scrape_feasible_now: validCategories.length >= minValidCategories,
    rows,
  };
}

File: llm_harness/stats/image/sources/arena_ai.py
"""Arena AI image source (native Python implementation)."""

from __future__ import annotations

import re
from typing import Any, TypedDict

from ...utils import (
    as_finite_number,
    fetch_with_timeout,
    now_epoch_seconds,
    percentile_rank,
)

ARENA_AI_BASE_URL = "https://arena.ai/leaderboard/text-to-image"
REQUEST_TIMEOUT_SECONDS = 30.0
MIN_VALID_ROWS = 20
MIN_VALID_CATEGORIES = 4
ARENA_AI_DEFAULT_CATEGORY_SLUGS = [
    "commercial-design",
    "3d-modeling",
    "cartoon",
    "photorealistic",
    "art",
    "portraits",
    "text-rendering",
]
ARENA_AI_GROUPED_CATEGORY_SLUGS = {
    "photorealistic": ["photorealistic", "portraits"],
    "illustrative": ["cartoon", "art"],
    "contextual": ["commercial-design", "3d-modeling", "text-rendering"],
}
ARENA_AI_GROUP_BY_SLUG = {slug: group_name for group_name, slugs in ARENA_AI_GROUPED_CATEGORY_SLUGS.items() for slug in slugs}


class ArenaAiImageOptions(TypedDict, total=False):
    category_slugs: list[str]
    min_valid_rows: int
    min_valid_categories: int


def _detect_challenge(html: str) -> bool:
    return bool(
        re.search(
            r"challenge-platform|__CF\$cv\$params|Verify you are human|Security Verification",
            html,
            re.IGNORECASE,
        )
    )


def _extract_leaderboard_rows(html: str) -> list[dict[str, Any]]:
    title_matches = list(re.finditer(r'<a[^>]*\btitle="([^"]+)"', html))
    rows: list[dict[str, Any]] = []
    for index, match in enumerate(title_matches):
        title = match.group(1) if match.group(1) else ""
        start = match.start()
        end = title_matches[index + 1].start() if index + 1 < len(title_matches) else min(start + 4000, len(html))
        row_block = html[start:end]
        score_match = re.search(r">(\d{3,4})</span><span[^>]*>±(\d+)</span>", row_block)
        votes_match = re.search(r"±\d+</span>(?:.|\n){0,1200}?>([\d,]{2,})</span>", row_block)
        provider_match = re.search(r">([^<]+ · [^<]+)</span>", row_block)
        rows.append(
            {
                "model": title,
                "provider": provider_match.group(1) if provider_match else None,
                "score": as_finite_number(score_match.group(1) if score_match else None),
                "ci95": f"±{score_match.group(2)}" if score_match else None,
                "votes": as_finite_number(votes_match.group(1).replace(",", "")) if votes_match else None,
            }
        )
    return rows


def _fetch_category(base_url: str, category_slug: str) -> dict[str, Any]:
    source_url = f"{base_url}/{category_slug}"
    try:
        response = fetch_with_timeout(
            source_url,
            headers={
                "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Safari/537.36",
                "accept-language": "en-US,en;q=0.9",
            },
            timeout_seconds=REQUEST_TIMEOUT_SECONDS,
        )
        html = response.text
        rows_with_score = [row for row in _extract_leaderboard_rows(html) if row["score"] is not None]
        return {
            "fetched_at_epoch_seconds": now_epoch_seconds(),
            "category_slug": category_slug,
            "source_url": source_url,
            "final_url": str(response.url),
            "status_code": response.status_code,
            "challenge_detected": _detect_challenge(html),
            "rows_with_score": len(rows_with_score),
            "rows": rows_with_score,
        }
    except Exception:
        return {
            "fetched_at_epoch_seconds": now_epoch_seconds(),
            "category_slug": category_slug,
            "source_url": source_url,
            "final_url": None,
            "status_code": None,
            "challenge_detected": False,
            "rows_with_score": 0,
            "rows": [],
        }


def _round4(value: float) -> float:
    return round(value, 4)


def _weighted_score_or_average(weighted_sum: float, votes_sum: float, score_sum: float, count: int) -> float | None:
    if votes_sum > 0:
        return _round4(weighted_sum / votes_sum)
    if count > 0:
        return _round4(score_sum / count)
    return None


def _build_grouped_scores(category_rows: dict[str, dict[str, Any]]) -> dict[str, Any]:
    grouped = {
        "photorealistic": {"score_weighted_sum": 0.0, "votes_sum": 0.0, "score_sum": 0.0, "count": 0},
        "illustrative": {"score_weighted_sum": 0.0, "votes_sum": 0.0, "score_sum": 0.0, "count": 0},
        "contextual": {"score_weighted_sum": 0.0, "votes_sum": 0.0, "score_sum": 0.0, "count": 0},
    }
    for category_slug, row in category_rows.items():
        group_name = ARENA_AI_GROUP_BY_SLUG.get(category_slug)
        if not group_name:
            continue
        score = row.get("score") or 0
        votes = row.get("votes") or 0
        grouped[group_name]["score_weighted_sum"] += score * votes
        grouped[group_name]["votes_sum"] += votes
        grouped[group_name]["score_sum"] += score
        grouped[group_name]["count"] += 1
    grouped_total_votes = sum(bucket["votes_sum"] for bucket in grouped.values())
    weighted_scores = {
        "photorealistic": _weighted_score_or_average(
            grouped["photorealistic"]["score_weighted_sum"],
            grouped["photorealistic"]["votes_sum"],
            grouped["photorealistic"]["score_sum"],
            grouped["photorealistic"]["count"],
        ),
        "illustrative": _weighted_score_or_average(
            grouped["illustrative"]["score_weighted_sum"],
            grouped["illustrative"]["votes_sum"],
            grouped["illustrative"]["score_sum"],
            grouped["illustrative"]["count"],
        ),
        "contextual": _weighted_score_or_average(
            grouped["contextual"]["score_weighted_sum"],
            grouped["contextual"]["votes_sum"],
            grouped["contextual"]["score_sum"],
            grouped["contextual"]["count"],
        ),
        "grouped_overall": _weighted_score_or_average(
            sum(bucket["score_weighted_sum"] for bucket in grouped.values()),
            grouped_total_votes,
            sum(bucket["score_sum"] for bucket in grouped.values()),
            sum(bucket["count"] for bucket in grouped.values()),
        ),
    }
    return {
        "weighted_scores": weighted_scores,
        "grouped_votes": {
            "photorealistic": int(grouped["photorealistic"]["votes_sum"]),
            "illustrative": int(grouped["illustrative"]["votes_sum"]),
            "contextual": int(grouped["contextual"]["votes_sum"]),
            "total": int(grouped_total_votes),
        },
    }


def _build_aggregated_rows(category_payloads: list[dict[str, Any]]) -> list[dict[str, Any]]:
    by_model: dict[str, dict[str, Any]] = {}
    for payload in category_payloads:
        for index, row in enumerate(payload.get("rows", [])):
            if not isinstance(row, dict):
                continue
            aggregate = by_model.setdefault(
                row["model"],
                {
                    "model": row["model"],
                    "provider": row.get("provider"),
                    "category_rows": {},
                    "category_count": 0,
                    "votes_sum": 0.0,
                    "score_sum": 0.0,
                    "score_weighted_sum": 0.0,
                    "rank_sum": 0.0,
                },
            )
            aggregate["category_rows"][payload["category_slug"]] = {
                "rank": index + 1,
                "score": row.get("score"),
                "ci95": row.get("ci95"),
                "votes": row.get("votes"),
            }
            aggregate["category_count"] += 1
            aggregate["rank_sum"] += index + 1
            aggregate["score_sum"] += row.get("score") or 0
            votes = row.get("votes") or 0
            aggregate["votes_sum"] += votes
            aggregate["score_weighted_sum"] += (row.get("score") or 0) * votes
    aggregated_rows: list[dict[str, Any]] = []
    for aggregate in by_model.values():
        average_score = _round4(aggregate["score_sum"] / aggregate["category_count"]) if aggregate["category_count"] > 0 else None
        vote_weighted_score = _round4(aggregate["score_weighted_sum"] / aggregate["votes_sum"]) if aggregate["votes_sum"] > 0 else average_score
        average_rank = _round4(aggregate["rank_sum"] / aggregate["category_count"]) if aggregate["category_count"] > 0 else None
        grouped = _build_grouped_scores(aggregate["category_rows"])
        aggregated_rows.append(
            {
                "model": aggregate["model"],
                "provider": aggregate["provider"],
                "category_count": aggregate["category_count"],
                "average_score": average_score,
                "vote_weighted_score": vote_weighted_score,
                "average_rank": average_rank,
                "votes_sum": int(aggregate["votes_sum"]),
                "categories": aggregate["category_rows"],
                "weighted_scores": grouped["weighted_scores"],
                "grouped_votes": grouped["grouped_votes"],
                "percentiles": {
                    "vote_weighted_percentile": None,
                    "photorealistic_percentile": None,
                    "illustrative_percentile": None,
                    "contextual_percentile": None,
                    "grouped_overall_percentile": None,
                },
            }
        )
    aggregated_rows.sort(
        key=lambda row: (
            row["vote_weighted_score"] if row["vote_weighted_score"] is not None else float("-inf"),
            row["category_count"],
        ),
        reverse=True,
    )
    vote_weighted_values = [row["vote_weighted_score"] for row in aggregated_rows]
    photorealistic_values = [row["weighted_scores"]["photorealistic"] for row in aggregated_rows]
    illustrative_values = [row["weighted_scores"]["illustrative"] for row in aggregated_rows]
    contextual_values = [row["weighted_scores"]["contextual"] for row in aggregated_rows]
    grouped_overall_values = [row["weighted_scores"]["grouped_overall"] for row in aggregated_rows]
    return [
        {
            **row,
            "percentiles": {
                "vote_weighted_percentile": percentile_rank(vote_weighted_values, row["vote_weighted_score"]),
                "photorealistic_percentile": percentile_rank(photorealistic_values, row["weighted_scores"]["photorealistic"]),
                "illustrative_percentile": percentile_rank(illustrative_values, row["weighted_scores"]["illustrative"]),
                "contextual_percentile": percentile_rank(contextual_values, row["weighted_scores"]["contextual"]),
                "grouped_overall_percentile": percentile_rank(grouped_overall_values, row["weighted_scores"]["grouped_overall"]),
            },
        }
        for row in aggregated_rows
    ]


def get_arena_ai_image_stats(
    options: ArenaAiImageOptions | None = None,
) -> dict[str, Any]:
    """Fetch and aggregate Arena text-to-image leaderboard categories."""
    options = options or {}
    category_slugs = options.get("category_slugs") or ARENA_AI_DEFAULT_CATEGORY_SLUGS
    min_valid_rows = options.get("min_valid_rows", MIN_VALID_ROWS)
    min_valid_categories = options.get("min_valid_categories", MIN_VALID_CATEGORIES)
    categories = [_fetch_category(ARENA_AI_BASE_URL, slug) for slug in category_slugs]
    valid_categories = [category for category in categories if category["rows_with_score"] >= min_valid_rows]
    rows = _build_aggregated_rows(valid_categories)
    return {
        "fetched_at_epoch_seconds": now_epoch_seconds(),
        "base_url": ARENA_AI_BASE_URL,
        "category_slugs": list(category_slugs),
        "categories": categories,
        "grouped_category_slugs": ARENA_AI_GROUPED_CATEGORY_SLUGS,
        "valid_categories": [category["category_slug"] for category in valid_categories],
        "total_valid_categories": len(valid_categories),
        "total_models_aggregated": len(rows),
        "scrape_feasible_now": len(valid_categories) >= min_valid_categories,
        "rows": rows,
    }
